x-pack/plugins/integration_assistant/server/graphs/ecs/graph.ts

```diff
@@ -135,38 +47,52 @@ function chainRouter(state: EcsMappingState): string {
     return 'invalidEcsFields';
   }
   if (!state.finalized) {
-    return 'modelOutput';
+    return 'modelSubOutput';
   }
   return END;
 }
 
-export async function getEcsGraph(model: ActionsClientChatOpenAI | ActionsClientSimpleChatModel) {
+// This is added as a separate graph to be able to run these steps concurrently from handleCreateMappingChunks
+async function getEcsSubGraph(model: ActionsClientChatOpenAI | ActionsClientSimpleChatModel) {
   const workflow = new StateGraph({
     channels: graphState,
   })
-    .addNode('modelInput', modelInput)
-    .addNode('modelOutput', modelOutput)
-    .addNode('handleEcsMapping', (state: EcsMappingState) => handleEcsMapping(state, model))
+    .addNode('modelSubOutput', modelSubOutput)
     .addNode('handleValidation', handleValidateMappings)
+    .addNode('handleEcsMapping', (state: EcsMappingState) => handleEcsMapping(state, model))
     .addNode('handleDuplicates', (state: EcsMappingState) => handleDuplicates(state, model))
     .addNode('handleMissingKeys', (state: EcsMappingState) => handleMissingKeys(state, model))
     .addNode('handleInvalidEcs', (state: EcsMappingState) => handleInvalidEcs(state, model))
-    .addEdge(START, 'modelInput')
-    .addEdge('modelOutput', END)
+    .addEdge(START, 'handleEcsMapping')
     .addEdge('handleEcsMapping', 'handleValidation')
     .addEdge('handleDuplicates', 'handleValidation')
     .addEdge('handleMissingKeys', 'handleValidation')
     .addEdge('handleInvalidEcs', 'handleValidation')
-    .addConditionalEdges('modelInput', inputRouter, {
-      ecsMapping: 'handleEcsMapping',
-      modelOutput: 'modelOutput',
-    })
     .addConditionalEdges('handleValidation', chainRouter, {
       duplicateFields: 'handleDuplicates',
       missingKeys: 'handleMissingKeys',
       invalidEcsFields: 'handleInvalidEcs',
-      modelOutput: 'modelOutput',
-    });
+      modelSubOutput: 'modelSubOutput',
+    })
+    .addEdge('modelSubOutput', END);
+
+  const compiledEcsGraph = workflow.compile();
+
+  return compiledEcsGraph;
+}
+
+export async function getEcsGraph(model: ActionsClientChatOpenAI | ActionsClientSimpleChatModel) {
+  const subGraph = await getEcsSubGraph(model);
+  const workflow = new StateGraph({
+    channels: graphState,
+  })
+    .addNode('modelInput', modelInput)
+    .addNode('modelOutput', modelOutput)
+    .addNode('subGraph', subGraph)
+    .addEdge(START, 'modelInput')
+    .addEdge('subGraph', 'modelOutput')
+    .addConditionalEdges('modelInput', handleCreateMappingChunks)
```

Comment:
Why is this conditionalEdge when it is always going to same node? Does Langgraph support `Send` calls only through ConditionalEdges?

Reply:
Indeed conditionalEdge is a requirement, The handleCreateMappingChunks is not actually a node. Its defined as a const with an anonymous async function, it returns the name of the node it should go to next I believe, once its done. Its why `subGraph` has an edge to `modelOutput.`
Example on the API reference: https://langchain-ai.github.io/langgraphjs/reference/classes/index.Send.html

Reply:
Indeed mentioned here - https://langchain-ai.github.io/langgraph/concepts/low_level/#send

`To support this design pattern, LangGraph supports returning [Send](https://langchain-ai.github.io/langgraph/reference/graphs/#send) objects from conditional edges. Send takes two arguments: first is the name of the node, and second is the state to pass to that node.`